src/main/clojure/clara/rules/memory.cljc

```diff
@@ -241,7 +256,78 @@
+                     ;; pre-1.9.0.
+                     (if-let [^clojure.lang.ISeq entries (.seq ^clojure.lang.Seqable bindings)]
+                       ;; Type hint to Indexed vs MapEntry just because MapEntry seems to be a
+                       ;; less stable impl detail to rely on.
+                       (loop [^clojure.lang.Indexed entry (.first entries)
+                              entries (.next entries)]
+                         (let [k (some-> entry (.nth 0))
+                               v (some-> entry (.nth 1))]
+                           (if (and k
+                                    ;; other-bindings will always be persistent map so invoke
+                                    ;; it directly.  It is faster than `get`.
+                                    (compare-fn v (other-bindings k)))
+                             (recur (some-> entries .first)
+                                    (some-> entries .next))
+                             ;; If there is no k left, then every entry matched.  If there is a k,
+                             ;; that means the comparison failed, so the maps aren't equal.
+                             (not k))))
+                       ;; Empty bindings on both sides.
+                       true))) 
+
+              ;; Check the :matches on each token.  :matches need to be in the same order on both
+              ;; tokens to be considered the same.
+              (let [^clojure.lang.Indexed matches (:matches token)
+                    ^clojure.lang.Indexed other-matches (:matches other-token)
+                    count-matches (.size ^java.util.List matches)]
+                (and (= count-matches
+                        (.size ^java.util.List other-matches))
+                     (loop [i 0]
+                       (cond
+                         (= i count-matches)
+                         true
+
+                         ;; Compare node-id's first.  Fallback to comparing facts.  This will
+                         ;; most very likely be the most expensive part to execute.
+                         (let [^clojure.lang.Indexed m (.nth matches i)
+                               ^clojure.lang.Indexed om (.nth other-matches i)]
+                           ;; A token :matches tuple is of the form [fact node-id].
+                           (and (= (.nth m 1) (.nth om 1))
+                                (compare-fn (.nth m 0) (.nth om 0))))
+                         (recur (inc i))
+
+                         :else
+                         false))))))))
+
+#?(:clj
+   (defprotocol IdentityComparable
+     (using-token-identity [this bool])))
+
+#?(:clj
+   (deftype RuleOrderedActivation [node-id
+                                   token
+                                   activation
+                                   rule-load-order
+                                   ^:unsynchronized-mutable use-token-identity?]
```

Comment:
I see what you're saying and I think this whole situation here is actually pretty sneaky.
I gave these a mutable flag to avoid more object allocation down in `remove-activations!`.

If you look carefully at [1] only local instances of RuleOrderedActivation are ever created that would mutate this field at all.  This means only a single thread would ever have one of these at a time.  The object is then discarded entirely.

`add-activations!` adds the only RuleOrderedActivation instances that would be able to be seen by multiple threads.  And those RuleOrderedActivation are never changed by having `using-token-identity` called on them.  They would always have this `use-token-identity?` field set to false (the default in `->rule-ordered-activation`).

The only reason the [2] RuleOrderedActivation.equals() implementation looks at this field on both `this` and `other` here is that I didn't want to make any assumptions about what object equals() was called via the PriorityQueue we are using.

I could have used a different object type to represent the "mutable" - strictly only local object types and the one that is shared with other threads later in memory.  However, I thought that'd also be clunky anyways and just more code.

This certainly needs some comments, I agree.  I can work on adding a few.

[1] https://github.com/rbrush/clara-rules/pull/214/files#diff-79e3c7ecc366c352f808f2a796662a30R806
[2] https://github.com/rbrush/clara-rules/pull/214/files#diff-79e3c7ecc366c352f808f2a796662a30R352